### ISSUE TITLE
Request for Clarification on Data Preparation for Reproducing NeCA Results

### ISSUE THREAD

**Issue by PenguinHater** [WangStephen/NeCA#1]
Hello,

I am currently attempting to reproduce the results from “NeCA: Coronary Artery Tree Reconstruction From Two 2D Projections Via Neural Implicit Representation” using the publicly available NeCA repository and the ImageCAS dataset. I am following the repository instructions and the descriptions provided in the paper as closely as possible.

According to the NeCA README, the expected input format is (1, Number of projections, Height, Width), which suggests that the model is trained using 2D projection images. However, during correspondence with the authors, I was advised to use the DeepCA preprocessing pipeline (specifically the CCTA_split function) to generate segmented 3D coronary vessel volumes (e.g., (128, 128, 128)), and that these segmented volumes serve as the ground truth data for NeCA.

This seems to imply an intermediate preprocessing step that is not explicitly documented in the NeCA repository or paper, namely the generation of 2D projections from the segmented 3D vessel volumes. At present, it is unclear how this projection step was performed in the original experiments. In particular, the projection geometry (e.g., orthogonal projections, MIP, ray-sum), the chosen view angles, and any normalization or scaling applied to the projections are not specified.

When projections are generated directly from the raw ImageCAS CTA volumes, the reproduced quantitative results differ significantly from those reported in the paper, which suggests that NeCA may rely on projections rendered from pre-segmented vessel geometry rather than raw CT data. Clarifying this step would greatly improve reproducibility and help align new experiments with the intended pipeline.

Could you please clarify, even at a high level, how the 2D projections used for NeCA training were generated from the segmented 3D vessel volumes? Any guidance on the projection method, camera setup, or normalization would be extremely helpful.

Thank you very much for making the code and data publicly available, and for your time in helping improve the reproducibility of this work.

Best regards,
George

**Comment by WangStephen** [comment]
Hi George,

Thank you so much for your interests in our work.

The preprocessing step is indicated in the paper section 2.1 including the table 1 of the projection geometry we used to generate the projections from the 3D coronary vessels. The important point here is that we used ODL to generate the projections so there would no domain gap to our NeCA method where we also use ODL as the differentiable projector layer.

The NeCA doesn't support raw CT reconstruction from only two projections. Based on our experiments, two projections work for segmented coronary vessel reconstruction. However, I think the NeCA pipeline can support raw CT reconstruction if you increase the number of projections. In my opinion, probably you can try 60, 90, 120 projections for raw CT reconstruction.

Regards,
Yiying

**Comment by PenguinHater** [comment]
Dear Dr. Wang,

Thank you very much for your response, and I apologize for the delayed reply.

I am continuing to use the 3D binarized segmented coronary vessel data you previously shared. Despite following the preprocessing and training steps described in the paper and repository, I am still unable to reproduce results consistent with those reported, and I was hoping to get your feedback on my process and observations. Below, I outline my full workflow and the issues I encountered.

Process followed:

I generated two projections from the provided 1.npy file using ODL, ensuring the geometry matched the parameters specified in the configuration file. The resulting projection tensor has dimensions (1, 2, width, height). As a sanity check, I also visualized the generated 2D projections in TensorBoard.

<img width="1292" height="637" alt="Image" src="https://github.com/user-attachments/assets/3e06a3bc-6bb5-4cdd-98b6-7f43dd6c9f50" />

I then ran train.py using these projections and allowed the script to complete fully. The training finished successfully after approximately two hours, and logging messages were printed as expected.

<img width="1247" height="230" alt="Image" src="https://github.com/user-attachments/assets/7219772f-082c-4f99-ab18-2a14fc44b43d" />

When analyzing the TensorBoard outputs, I noticed that the generated events.out.tfevents file only contained a single scalar labeled "train". Unlike my experience with DeepCA, no additional metrics appeared to be logged.

<img width="1243" height="296" alt="Image" src="https://github.com/user-attachments/assets/7cf2daff-de65-49f4-85c5-40ef8262ec45" />

To further evaluate performance, I wrote a separate script to compute the Dice score and CD (L2) distance from the saved checkpoint outputs. Both metrics showed large discrepancies compared to the values reported in the original paper.

I also generated multiple cross-sectional slices of the reconstructed volume from different viewing angles. These confirmed that the reconstructed output did not overlap well with the ground truth.

<img width="1248" height="566" alt="Image" src="https://github.com/user-attachments/assets/7fc02b5e-f72e-42c3-9db4-706b019f50ff" />

To better understand the failure mode, I generated 3D visualizations of both the reconstructed volume and the ground truth as an additional sanity check. The ground truth appeared normal; however, the reconstructed volume formed an entirely different structure. When rotated, it appeared to resemble two “X”-shaped structures connected by a tube, rather than a coronary vessel.

<img width="689" height="530" alt="Image" src="https://github.com/user-attachments/assets/761dd154-18fa-4298-8a0e-70aa4b9ef36f" />

<img width="480" height="650" alt="Image" src="https://github.com/user-attachments/assets/f8729925-e316-4192-a448-925b1e8f4f02" />

I repeated the entire pipeline using a different set of projection angles ([0.01°, 0.0°] and [90.0°, 0.0°]). While parts of the vessel structure seemed partially recovered, the reconstruction was still significantly smaller than the ground truth, appeared rotated, and contained apparent outlier points in the background.

<img width="671" height="648" alt="Image" src="https://github.com/user-attachments/assets/c0a86913-88fc-429e-9da8-a390f2740049" />

Questions:

Are the default configuration file values sufficient for proper reconstruction on the ImageCAS dataset? If not, would it be possible for you to share the exact configuration parameters used for those experiments?

Could you clarify how training results and intermediate outputs are intended to be accessed? Specifically, should quantitative metrics or visualizations appear in TensorBoard, or are they primarily stored in the saved epoch/checkpoint files? I may be misunderstanding how the logging is intended to work, as the README only mentions that outputs are saved under ./logs/.

Thank you very much for your time and for making your work publicly available. I greatly appreciate any guidance you can provide.

Best regards,
George

**Comment by WangStephen** [comment]
Hi George,

Happy New Year.

1. The default configuration file values are sufficient for the proper reconstruction.

There is one problem of your generated projections, which is kinda strange from what it should look like. My two projections for RCA are similar to Figure 1 in the paper.

Could you please check how you generate the two projections? ODL geometry actually is a little counterintuitive. Please check my code (https://github.com/WangStephen/NeCA/blob/77246df0826a031c12072f6ea64bb3370a0baf6b/src/trainer.py#L72C7-L101C9) to ensure you generate two projections following the same protocol as what it is in the code.

2. Yes, I do not save evaluation values such as Dice and something else to the tensorboard in the code, because the code is for data when there is no 3D ground truth. If you want to have these results for testing purpose, you can directly add them in (https://github.com/WangStephen/NeCA/blob/77246df0826a031c12072f6ea64bb3370a0baf6b/src/trainer.py#L204C12-L207C1). Just remember to load your ground truth data and measure the results with the current prediction, and then save to tensorboard.

Hope these help.

regards,
Yiying

**Comment by PenguinHater** [comment]
Dear Dr. Wang,

I created a standalone script that follows the same projection-generation steps described in the protocol you shared; however, the NeCA results still do not appear to be valid.

Specifically, I replicated the projection process by:

Reordering the angles to convert [pitch, yaw] into [yaw, pitch]

Defining the same reference DSO vectors

Applying the yaw rotation first

Applying the pitch rotation second

Converting the resulting rotation to an axis–angle representation for ODL

Constructing the ConeBeam3DProjector

This procedure was applied identically for both projections. The generated 2D projections themselves appear reasonable based on visual inspection:
<img width="958" height="367" alt="Image" src="https://github.com/user-attachments/assets/b7178ebd-d52d-4878-a986-e397b5b57fb2" />

However, when reconstructing and visualizing the 3D result, the output still exhibits an unexpected and distorted shape:
<img width="947" height="379" alt="Image" src="https://github.com/user-attachments/assets/a16aafe4-be10-4112-86e9-2f44aca4e505" />

For reference, the exact code used to generate these projections is available here:
https://github.com/PenguinHater/NeCA-Projection-Generation-Script/blob/main/generate_neca_projections_faithful.py

I would greatly appreciate any guidance on what might be causing this discrepancy, as I am unsure where the issue arises.

Thank you for your time and help.

Best regards,
George

**Comment by WangStephen** [comment]
Hi George,

I found one main problem:
In the line 110-111 in your code, the dimension of the 3D data to the projection function is [1,1,128,128,128], but it should be [1,128,128,128].

Also try to delete 118-120, the normalisation steps initially to reduce other factors. Once you can successfully reconstruct, then you can add the normalisation steps to see if it still works.

regards,
Yiying

**Comment by PenguinHater** [comment]
Hello Dr. Wang,

Everything appears to be working now, thank you very much for your help. I tested the Dice and CD(l2) scores with the results, and the metrics now match the ones reported in the paper. I have also updated my Github script incase others wish to replicate the results.

<img width="1602" height="467" alt="Image" src="https://github.com/user-attachments/assets/bbc5e3f1-f40f-4cf3-b2ff-2a5d332586f2" />

Thanks again for your patience and guidance,
George

**Comment by WangStephen** [comment]
Hi George,

Happy to see it works now.

There is one more improvement suggestion as instructed in the last sentence of the section 2.5 of the paper, i.e., applying connected component analysis. This may improve just a little to RCA results but a lot to the LAD results as the initial LAD reconstruction from NeCA would have many outliers.

regards,
Yiying
